File: taurus-protect-sdk-go/pkg/protect/model/rule_cell.go
package model

import "math/big"

// RuleCell is a typed transaction-rule cell. Each concrete type corresponds to a
// column type and one of its cell-type enum values in the governance protobuf
// schema (e.g. FiatAmountRange is the RuleFiatAmountRange cell of a
// RuleFiatAmount column).
//
// The *Any cell types are cells like any other. Being the protobuf zero
// values, their serialized form happens to be the empty cell. A nil RuleCell
// is treated as an empty cell at encode; decode yields nil only for cells
// under columns that have no cell types at all (RuleAny or unknown columns).
//
// Cells decoded from a container built with a newer schema than this SDK
// (unknown cell type, or unknown protobuf fields inside the cell) are
// represented as RawCell and re-encoded verbatim, so no data is ever silently
// dropped. See RawCell.
type RuleCell interface {
	isRuleCell()

	// Kind is the cell's discriminator — the variant name, e.g. "SourceInternalWallet"
	// or "RawCell". Java kind(), Python .kind and TS .kind all expose the same string,
	// so a caller can branch or log without a Go type switch and without a per-language
	// mapping table.
	Kind() string
}

// RawCell preserves a cell whose type or content is unknown to this SDK
// version. Its bytes are re-emitted verbatim on encode.
type RawCell struct {
	// ColumnType is the column type string the cell was decoded under
	// (may itself be a numeric passthrough for unknown column types).
	ColumnType string
	// Payload is the exact wire content of the cell. Named to match Java, Python
	// and TS, which all call it payload.
	Payload []byte
}

// --- RuleFiatAmount column ---

// FiatAmountAny matches any fiat amount.
type FiatAmountAny struct{}

// FiatAmountIsZero matches a zero fiat amount.
type FiatAmountIsZero struct{}

// FiatAmountRange matches a fiat amount within [MinAmount, MaxAmount].
type FiatAmountRange struct {
	// MinAmount is the inclusive lower bound as a decimal string.
	MinAmount string
	// MaxAmount is the inclusive upper bound as a decimal string.
	MaxAmount string
}

// --- RuleSource column ---

// SourceAny matches any transaction source.
type SourceAny struct{}

// SourceInternalWallet matches an internal wallet source by derivation path.
type SourceInternalWallet struct {
	Path string
}

// SourceInternalAddress matches an internal address source.
type SourceInternalAddress struct {
	Address string
	Path    string
}

// SourceAnyExchange matches any exchange source.
type SourceAnyExchange struct{}

// SourceExchange matches a specific exchange source by label.
type SourceExchange struct {
	Label string
}

// SourceExternalAddress matches an external address source.
type SourceExternalAddress struct {
	Address string
	Memo    string
}

// --- RuleDestination column ---

// DestinationAny matches any transaction destination.
type DestinationAny struct{}

// DestinationInternalWallet matches an internal wallet destination by derivation path.
type DestinationInternalWallet struct {
	Path string
}

// DestinationInternalAddress matches an internal address destination.
type DestinationInternalAddress struct {
	Address string
	Path    string
}

// DestinationExternalAddress matches an external (whitelisted) address destination.
type DestinationExternalAddress struct {
	Address string
	Memo    string
}

// DestinationAnyExchange matches any exchange destination.
type DestinationAnyExchange struct{}

// DestinationExchange matches a specific exchange destination.
type DestinationExchange struct {
	Label string
	Memo  string
}

// DestinationContractAddress matches a specific contract address destination.
type DestinationContractAddress struct {
	Address string
	Name    string
	Symbol  string
	// Blockchain is the blockchain enum value name (e.g. "ETH").
	Blockchain string
}

// DestinationAnyExternalAddress matches any external address destination.
type DestinationAnyExternalAddress struct{}

// DestinationAnyContractAddress matches any contract address destination.
type DestinationAnyContractAddress struct{}

// --- RuleStringEqual column (payload is the raw UTF-8 string bytes) ---

// StringEqualAny matches any string value.
type StringEqualAny struct{}

// StringEqualEmpty matches an empty string value.
type StringEqualEmpty struct{}

// StringEqualValue matches a specific string value.
type StringEqualValue struct {
	Value string
}

// --- RuleBytesEqual column (payload is the raw bytes) ---

// BytesEqualAny matches any bytes value.
type BytesEqualAny struct{}

// BytesEqualEmpty matches an empty bytes value.
type BytesEqualEmpty struct{}

// BytesEqualValue matches a specific bytes value.
type BytesEqualValue struct {
	Value []byte
}

// --- RuleStringArrayEqual column ---

// StringArrayEqualAny matches any string array.
type StringArrayEqualAny struct{}

// StringArrayEqualEmpty matches an empty string array.
type StringArrayEqualEmpty struct{}

// StringArrayEqualValue matches a specific string array.
type StringArrayEqualValue struct {
	Values []string
}

// --- RuleIntegerGreater column ---

// IntegerGreaterAny matches any signed integer.
type IntegerGreaterAny struct{}

// IntegerGreaterValue matches signed integers greater than Value.
//
// The wire format carries the magnitude with the sign expressed by the cell
// type enum (RuleIntegerGreaterValue vs RuleIntegerGreaterNegValue); the codec
// selects the arm from Value's sign, so callers just set the signed value.
type IntegerGreaterValue struct {
	Value *big.Int
}

// --- RuleUIntegerGreater column ---

// UIntegerGreaterAny matches any unsigned integer.
type UIntegerGreaterAny struct{}

// UIntegerGreaterIsZero matches a zero unsigned integer.
type UIntegerGreaterIsZero struct{}

// UIntegerGreaterValue matches unsigned integers greater than Value (must be non-negative).
type UIntegerGreaterValue struct {
	Value *big.Int
}

// UIntegerGreaterIsEqual matches unsigned integers equal to Value (must be non-negative).
type UIntegerGreaterIsEqual struct {
	Value *big.Int
}

// --- RuleWhitelistedContract column ---

// WhitelistedContractAny matches any whitelisted contract.
type WhitelistedContractAny struct{}

// WhitelistedContractAddress matches a specific whitelisted contract address
// (the RuleWhitelistedContract_RuleDestinationContractAddress cell type).
type WhitelistedContractAddress struct {
	Address string
	Name    string
	Symbol  string
	// Blockchain is the blockchain enum value name (e.g. "ETH").
	Blockchain string
}

func (RawCell) isRuleCell()                       {}
func (FiatAmountAny) isRuleCell()                 {}
func (FiatAmountIsZero) isRuleCell()              {}
func (FiatAmountRange) isRuleCell()               {}
func (SourceAny) isRuleCell()                     {}
func (SourceInternalWallet) isRuleCell()          {}
func (SourceInternalAddress) isRuleCell()         {}
func (SourceAnyExchange) isRuleCell()             {}
func (SourceExchange) isRuleCell()                {}
func (SourceExternalAddress) isRuleCell()         {}
func (DestinationAny) isRuleCell()                {}
func (DestinationInternalWallet) isRuleCell()     {}
func (DestinationInternalAddress) isRuleCell()    {}
func (DestinationExternalAddress) isRuleCell()    {}
func (DestinationAnyExchange) isRuleCell()        {}
func (DestinationExchange) isRuleCell()           {}
func (DestinationContractAddress) isRuleCell()    {}
func (DestinationAnyExternalAddress) isRuleCell() {}
func (DestinationAnyContractAddress) isRuleCell() {}
func (StringEqualAny) isRuleCell()                {}
func (StringEqualEmpty) isRuleCell()              {}
func (StringEqualValue) isRuleCell()              {}
func (BytesEqualAny) isRuleCell()                 {}
func (BytesEqualEmpty) isRuleCell()               {}
func (BytesEqualValue) isRuleCell()               {}
func (StringArrayEqualAny) isRuleCell()           {}
func (StringArrayEqualEmpty) isRuleCell()         {}
func (StringArrayEqualValue) isRuleCell()         {}
func (IntegerGreaterAny) isRuleCell()             {}
func (IntegerGreaterValue) isRuleCell()           {}
func (UIntegerGreaterAny) isRuleCell()            {}
func (UIntegerGreaterIsZero) isRuleCell()         {}
func (UIntegerGreaterValue) isRuleCell()          {}
func (UIntegerGreaterIsEqual) isRuleCell()        {}
func (WhitelistedContractAny) isRuleCell()        {}
func (WhitelistedContractAddress) isRuleCell()    {}

// Kind returns each variant's discriminator string; see RuleCell.Kind.
func (RawCell) Kind() string                       { return "RawCell" }
func (FiatAmountAny) Kind() string                 { return "FiatAmountAny" }
func (FiatAmountIsZero) Kind() string              { return "FiatAmountIsZero" }
func (FiatAmountRange) Kind() string               { return "FiatAmountRange" }
func (SourceAny) Kind() string                     { return "SourceAny" }
func (SourceInternalWallet) Kind() string          { return "SourceInternalWallet" }
func (SourceInternalAddress) Kind() string         { return "SourceInternalAddress" }
func (SourceAnyExchange) Kind() string             { return "SourceAnyExchange" }
func (SourceExchange) Kind() string                { return "SourceExchange" }
func (SourceExternalAddress) Kind() string         { return "SourceExternalAddress" }
func (DestinationAny) Kind() string                { return "DestinationAny" }
func (DestinationInternalWallet) Kind() string     { return "DestinationInternalWallet" }
func (DestinationInternalAddress) Kind() string    { return "DestinationInternalAddress" }
func (DestinationExternalAddress) Kind() string    { return "DestinationExternalAddress" }
func (DestinationAnyExchange) Kind() string        { return "DestinationAnyExchange" }
func (DestinationExchange) Kind() string           { return "DestinationExchange" }
func (DestinationContractAddress) Kind() string    { return "DestinationContractAddress" }
func (DestinationAnyExternalAddress) Kind() string { return "DestinationAnyExternalAddress" }
func (DestinationAnyContractAddress) Kind() string { return "DestinationAnyContractAddress" }
func (StringEqualAny) Kind() string                { return "StringEqualAny" }
func (StringEqualEmpty) Kind() string              { return "StringEqualEmpty" }
func (StringEqualValue) Kind() string              { return "StringEqualValue" }
func (BytesEqualAny) Kind() string                 { return "BytesEqualAny" }
func (BytesEqualEmpty) Kind() string               { return "BytesEqualEmpty" }
func (BytesEqualValue) Kind() string               { return "BytesEqualValue" }
func (StringArrayEqualAny) Kind() string           { return "StringArrayEqualAny" }
func (StringArrayEqualEmpty) Kind() string         { return "StringArrayEqualEmpty" }
func (StringArrayEqualValue) Kind() string         { return "StringArrayEqualValue" }
func (IntegerGreaterAny) Kind() string             { return "IntegerGreaterAny" }
func (IntegerGreaterValue) Kind() string           { return "IntegerGreaterValue" }
func (UIntegerGreaterAny) Kind() string            { return "UIntegerGreaterAny" }
func (UIntegerGreaterIsZero) Kind() string         { return "UIntegerGreaterIsZero" }
func (UIntegerGreaterValue) Kind() string          { return "UIntegerGreaterValue" }
func (UIntegerGreaterIsEqual) Kind() string        { return "UIntegerGreaterIsEqual" }
func (WhitelistedContractAny) Kind() string        { return "WhitelistedContractAny" }
func (WhitelistedContractAddress) Kind() string    { return "WhitelistedContractAddress" }

File: taurus-protect-sdk-java/client/src/main/java/com/taurushq/sdk/protect/client/model/rulescontainer/RuleCell.java
package com.taurushq.sdk.protect.client.model.rulescontainer;

import java.math.BigInteger;
import java.util.List;

import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;
import org.apache.commons.lang3.builder.ToStringBuilder;

/**
 * Typed transaction-rule cell (cross-SDK contract).
 *
 * <p>Each concrete subclass corresponds to a column type and one of its
 * cell-type enum values in the governance protobuf schema (e.g.
 * {@link FiatAmountRange} is the {@code RuleFiatAmountRange} cell of a
 * {@code RuleFiatAmount} column). The {@code *Any} cells are the protobuf zero
 * values, so their serialized form is the empty cell. Cells whose type/content
 * this SDK version cannot represent losslessly are preserved verbatim as
 * {@link RawCell}, so nothing is ever silently dropped.
 *
 * <p>The same set exists in the Go, Python and TypeScript SDKs, and the wire
 * encoding is pinned by the shared golden vectors at
 * {@code scripts/resources/governance-cell-vectors.json}. Java 8 has no sealed
 * types; the closed set is enforced by convention and the shared vectors.
 *
 * <p>Encode/decode against a column's byte cells lives in {@link RuleCellCodec}.
 */
public abstract class RuleCell {

    /** The model type name (matches the other SDKs' discriminator). */
    public abstract String kind();

    @Override
    public boolean equals(Object o) {
        return EqualsBuilder.reflectionEquals(this, o, false);
    }

    @Override
    public int hashCode() {
        return HashCodeBuilder.reflectionHashCode(this);
    }

    @Override
    public String toString() {
        return ToStringBuilder.reflectionToString(this);
    }

    /** Shared base for cells that carry no fields (the {@code *Any}/{@code *IsZero} arms). */
    private abstract static class NoField extends RuleCell {
        NoField() {
            super();
        }
    }

    // --- RuleFiatAmount ---
    public static final class FiatAmountAny extends NoField {
        @Override
        public String kind() { return "FiatAmountAny"; }
    }

    public static final class FiatAmountIsZero extends NoField {
        @Override
        public String kind() { return "FiatAmountIsZero"; }
    }

    public static final class FiatAmountRange extends RuleCell {
        public final String minAmount;
        public final String maxAmount;
        public FiatAmountRange(String minAmount, String maxAmount) { this.minAmount = minAmount; this.maxAmount = maxAmount; }
        @Override
        public String kind() { return "FiatAmountRange"; }
    }

    // --- RuleSource ---
    public static final class SourceAny extends NoField {
        @Override
        public String kind() { return "SourceAny"; }
    }

    public static final class SourceInternalWallet extends RuleCell {
        public final String path;
        public SourceInternalWallet(String path) { this.path = path; }
        @Override
        public String kind() { return "SourceInternalWallet"; }
    }

    public static final class SourceInternalAddress extends RuleCell {
        public final String address;
        public final String path;
        public SourceInternalAddress(String address, String path) { this.address = address; this.path = path; }
        @Override
        public String kind() { return "SourceInternalAddress"; }
    }

    public static final class SourceAnyExchange extends NoField {
        @Override
        public String kind() { return "SourceAnyExchange"; }
    }

    public static final class SourceExchange extends RuleCell {
        public final String label;
        public SourceExchange(String label) { this.label = label; }
        @Override
        public String kind() { return "SourceExchange"; }
    }

    public static final class SourceExternalAddress extends RuleCell {
        public final String address;
        public final String memo;
        public SourceExternalAddress(String address, String memo) { this.address = address; this.memo = memo; }
        @Override
        public String kind() { return "SourceExternalAddress"; }
    }

    // --- RuleDestination ---
    public static final class DestinationAny extends NoField {
        @Override
        public String kind() { return "DestinationAny"; }
    }

    public static final class DestinationInternalWallet extends RuleCell {
        public final String path;
        public DestinationInternalWallet(String path) { this.path = path; }
        @Override
        public String kind() { return "DestinationInternalWallet"; }
    }

    public static final class DestinationInternalAddress extends RuleCell {
        public final String address;
        public final String path;
        public DestinationInternalAddress(String address, String path) { this.address = address; this.path = path; }
        @Override
        public String kind() { return "DestinationInternalAddress"; }
    }

    public static final class DestinationExternalAddress extends RuleCell {
        public final String address;
        public final String memo;
        public DestinationExternalAddress(String address, String memo) { this.address = address; this.memo = memo; }
        @Override
        public String kind() { return "DestinationExternalAddress"; }
    }

    public static final class DestinationAnyExchange extends NoField {
        @Override
        public String kind() { return "DestinationAnyExchange"; }
    }

    public static final class DestinationExchange extends RuleCell {
        public final String label;
        public final String memo;
        public DestinationExchange(String label, String memo) { this.label = label; this.memo = memo; }
        @Override
        public String kind() { return "DestinationExchange"; }
    }

    public static final class DestinationContractAddress extends RuleCell {
        public final String address;
        public final String name;
        public final String symbol;
        public final String blockchain;
        public DestinationContractAddress(String address, String name, String symbol, String blockchain) {
            this.address = address; this.name = name; this.symbol = symbol; this.blockchain = blockchain;
        }
        @Override
        public String kind() { return "DestinationContractAddress"; }
    }

    public static final class DestinationAnyExternalAddress extends NoField {
        @Override
        public String kind() { return "DestinationAnyExternalAddress"; }
    }

    public static final class DestinationAnyContractAddress extends NoField {
        @Override
        public String kind() { return "DestinationAnyContractAddress"; }
    }

    // --- RuleStringEqual (payload is raw UTF-8 string bytes) ---
    public static final class StringEqualAny extends NoField {
        @Override
        public String kind() { return "StringEqualAny"; }
    }

    public static final class StringEqualEmpty extends NoField {
        @Override
        public String kind() { return "StringEqualEmpty"; }
    }

    public static final class StringEqualValue extends RuleCell {
        public final String value;
        public StringEqualValue(String value) { this.value = value; }
        @Override
        public String kind() { return "StringEqualValue"; }
    }

    // --- RuleBytesEqual (payload is raw bytes) ---
    public static final class BytesEqualAny extends NoField {
        @Override
        public String kind() { return "BytesEqualAny"; }
    }

    public static final class BytesEqualEmpty extends NoField {
        @Override
        public String kind() { return "BytesEqualEmpty"; }
    }

    public static final class BytesEqualValue extends RuleCell {
        public final byte[] value;
        public BytesEqualValue(byte[] value) { this.value = value; }
        @Override
        public String kind() { return "BytesEqualValue"; }
    }

    // --- RuleStringArrayEqual ---
    public static final class StringArrayEqualAny extends NoField {
        @Override
        public String kind() { return "StringArrayEqualAny"; }
    }

    public static final class StringArrayEqualEmpty extends NoField {
        @Override
        public String kind() { return "StringArrayEqualEmpty"; }
    }

    public static final class StringArrayEqualValue extends RuleCell {
        public final List<String> values;
        public StringArrayEqualValue(List<String> values) { this.values = values; }
        @Override
        public String kind() { return "StringArrayEqualValue"; }
    }

    // --- RuleIntegerGreater (sign selects the Value/NegValue wire arm) ---
    public static final class IntegerGreaterAny extends NoField {
        @Override
        public String kind() { return "IntegerGreaterAny"; }
    }

    public static final class IntegerGreaterValue extends RuleCell {
        public final BigInteger value;
        public IntegerGreaterValue(BigInteger value) { this.value = value; }
        @Override
        public String kind() { return "IntegerGreaterValue"; }
    }

    // --- RuleUIntegerGreater ---
    public static final class UIntegerGreaterAny extends NoField {
        @Override
        public String kind() { return "UIntegerGreaterAny"; }
    }

    public static final class UIntegerGreaterIsZero extends NoField {
        @Override
        public String kind() { return "UIntegerGreaterIsZero"; }
    }

    public static final class UIntegerGreaterValue extends RuleCell {
        public final BigInteger value;
        public UIntegerGreaterValue(BigInteger value) { this.value = value; }
        @Override
        public String kind() { return "UIntegerGreaterValue"; }
    }

    public static final class UIntegerGreaterIsEqual extends RuleCell {
        public final BigInteger value;
        public UIntegerGreaterIsEqual(BigInteger value) { this.value = value; }
        @Override
        public String kind() { return "UIntegerGreaterIsEqual"; }
    }

    // --- RuleWhitelistedContract ---
    public static final class WhitelistedContractAny extends NoField {
        @Override
        public String kind() { return "WhitelistedContractAny"; }
    }

    public static final class WhitelistedContractAddress extends RuleCell {
        public final String address;
        public final String name;
        public final String symbol;
        public final String blockchain;
        public WhitelistedContractAddress(String address, String name, String symbol, String blockchain) {
            this.address = address; this.name = name; this.symbol = symbol; this.blockchain = blockchain;
        }
        @Override
        public String kind() { return "WhitelistedContractAddress"; }
    }

    // --- Fallback for cells unknown to this SDK version (preserved verbatim) ---
    public static final class RawCell extends RuleCell {
        public final String columnType;
        public final byte[] payload;
        public RawCell(String columnType, byte[] payload) { this.columnType = columnType; this.payload = payload; }
        @Override
        public String kind() { return "RawCell"; }
    }
}
